pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConnectionController.java

```diff
@@ -0,0 +1,127 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface ConnectionController {
+
+    /**
+     * Increase the number of connections counter.
+     * @param remoteAddress
+     * @return
+     */
+    boolean increaseConnection(SocketAddress remoteAddress);
+
+    /**
+     * Decrease the number of connections counter.
+     * @param remoteAddress
+     */
+    void decreaseConnection(SocketAddress remoteAddress);
+
+
+    class DefaultConnectionController implements ConnectionController {
+        private static final Logger log = LoggerFactory.getLogger(DefaultConnectionController.class);
+        private static final Map<String, MutableInt> CONNECTIONS = new HashMap<>();
+        private static final Pattern IPV4_PATTERN = Pattern
+                .compile("^" + "(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)" + "(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}" + "$");
+        private static final ReentrantLock lock = new ReentrantLock();
+        private static int totalConnectionNum = 0;
+
+        private final int maxConnections;
+        private final int maxConnectionPerIp;
+
+        public DefaultConnectionController(ServiceConfiguration configuration) {
+            this.maxConnections = configuration.getBrokerMaxConnections();
+            this.maxConnectionPerIp = configuration.getBrokerMaxConnectionsPerIp();
+        }
+
+        @Override
+        public boolean increaseConnection(SocketAddress remoteAddress) {
+            if (maxConnections <= 0 && maxConnectionPerIp <= 0) {
+                return true;
+            }
+            if (!(remoteAddress instanceof InetSocketAddress) || !IPV4_PATTERN.matcher(
+                    ((InetSocketAddress) remoteAddress).getHostString()).find()) {
+                return true;
+            }
+            lock.lock();
+            try {
+                String ip = ((InetSocketAddress) remoteAddress).getHostString();
+                if (maxConnectionPerIp > 0) {
+                    CONNECTIONS.computeIfAbsent(ip, (x) -> new MutableInt(0)).increment();
+                }
+                if (maxConnections > 0) {
```

Comment:
Thanks for reviewing.

`Also, it might be easer in the construction to add a flags indicating what is enabled. For example,
maxConnectionsLimitEnabled and maxConnectionsLimitByIpEnabled`
OK, I will use  enabled flag

`I think this complicates the logic and the connection counting should only be incremented after checking if it's less than max connections. Total connections should never go above max connections right?`
In the channelActive event, the connection has been established, why not increase the count first?
If we don't do this, the total number of connections has actually been exceeded, but we make it looks like has not .